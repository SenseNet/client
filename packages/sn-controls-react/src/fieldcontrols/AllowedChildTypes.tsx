import CircularProgress from '@material-ui/core/CircularProgress'
import ClickAwayListener from '@material-ui/core/ClickAwayListener'
import FormControl from '@material-ui/core/FormControl'
import FormHelperText from '@material-ui/core/FormHelperText'
import FormLabel from '@material-ui/core/FormLabel'
import IconButton from '@material-ui/core/IconButton'
import InputAdornment from '@material-ui/core/InputAdornment'
import List from '@material-ui/core/List'
import ListItem from '@material-ui/core/ListItem'
import ListItemIcon from '@material-ui/core/ListItemIcon'
import ListItemSecondaryAction from '@material-ui/core/ListItemSecondaryAction'
import ListItemText from '@material-ui/core/ListItemText'
import Paper from '@material-ui/core/Paper'
import TextField from '@material-ui/core/TextField'
import { ConstantContent, ODataCollectionResponse } from '@sensenet/client-core'
import { ContentType, GenericContent } from '@sensenet/default-content-types'
import React, { Component } from 'react'
import { typeicons } from '../assets/icons'
import { renderIconDefault } from './icon'
import { ReactClientFieldSetting } from './ClientFieldSetting'

const INPUT_PLACEHOLDER = 'Start typing to add another type'
const ITEM_HEIGHT = 48

const styles = {
  inputContainer: {
    padding: '2px 4px',
    display: 'flex',
    alignItems: 'center',
    boxShadow: 'none',
    position: 'relative',
  },
  input: {
    marginLeft: 8,
    flex: 1,
  },
  button: {
    padding: 10,
  },
  listContainer: {
    position: 'absolute',
    top: '40px',
    maxHeight: ITEM_HEIGHT * 2.5,
    overflow: 'auto',
    zIndex: 10,
  },
  ddIsOpened: {
    display: 'block',
  },
  ddIsClosed: {
    display: 'none',
  },
}

const compare = (a: GenericContent, b: GenericContent) => {
  if (a.Name < b.Name) {
    return -1
  }
  if (a.Name > b.Name) {
    return 1
  }
  return 0
}

/**
 * Interface for AllowedChildTypes state
 */
export interface AllowedChildTypesState {
  value: string[]
  effectiveAllowedChildTypes: ContentType[]
  allowedTypesOnCTD: ContentType[]
  items: ContentType[]
  removeable: boolean
  allCTDs: ContentType[]
  isLoading: boolean
  inputValue: string
  isOpened: boolean
  anchorEl: HTMLElement
  getMenuItem: (item: ContentType, select: (item: ContentType) => void) => JSX.Element
  filteredList: ContentType[]
  selected: ContentType | null
}
/**
 * Field control that represents an AllowedChildTypes field. Available values will be populated from the FieldSettings.
 */
export class AllowedChildTypes extends Component<ReactClientFieldSetting, AllowedChildTypesState> {
  constructor(props: AllowedChildTypes['props']) {
    super(props)
    this.state = {
      value: [],
      effectiveAllowedChildTypes: [],
      allowedTypesOnCTD: [],
      items: [],
      removeable: true,
      allCTDs: [],
      isLoading: false,
      inputValue: '',
      isOpened: false,
      anchorEl: null as any,
      getMenuItem: (item: ContentType, select: (item: ContentType) => void) => (
        <ListItem key={item.Id} value={item.Id} onClick={() => select(item)} style={{ margin: 0 }}>
          <ListItemIcon style={{ margin: 0 }}>
            {this.props.renderIcon
              ? this.props.renderIcon(item.Icon ? item.Icon.toLowerCase() : 'contenttype')
              : renderIconDefault(
                  item.Icon && typeicons[item.Icon.toLowerCase()]
                    ? typeicons[item.Icon.toLowerCase()]
                    : typeicons.contenttype,
                )}
          </ListItemIcon>
          <ListItemText primary={item.DisplayName} />
        </ListItem>
      ),
      filteredList: [],
      selected: null,
    }
    this.handleSelect = this.handleSelect.bind(this)
    this.handleClickAway = this.handleClickAway.bind(this)
    this.getAllowedChildTypes()
    this.getAllContentTypes()
  }

  /**
   * component will unmount
   */
  public componentWillUnmount() {
    this.willUnmount = true
  }

  private willUnmount: boolean = false

  private async getAllowedChildTypes() {
    if (!this.props.repository) {
      throw new Error('You must pass a repository to this control')
    }
    if (!this.props.content) {
      return
    }

    const result = await this.props.repository.load<GenericContent>({
      idOrPath: this.props.content.Id,
      oDataOptions: {
        select: 'EffectiveAllowedChildTypes',
        expand: 'EffectiveAllowedChildTypes',
      },
    })
    if (this.willUnmount) {
      return
    }

    const allowedChildTypesFromCTD = (await this.props.repository.executeAction({
      idOrPath: this.props.content.Id,
      name: 'GetAllowedChildTypesFromCTD',
      method: 'GET',
      body: {
        select: ['Name', 'DisplayName', 'Icon'],
      },
    })) as ODataCollectionResponse<ContentType>

    const typeResults = result.d.EffectiveAllowedChildTypes as ContentType[]

    const types = this.getTypes(typeResults, allowedChildTypesFromCTD)

    this.setState({
      effectiveAllowedChildTypes: typeResults,
      items: types,
      removeable: typeResults.length === 0 || this.props.actionName !== 'new',
      value: types.map((t: ContentType) => t.Name),
    })
  }

  private getTypes(typeResults: ContentType[], allowedChildTypesFromCTD: ODataCollectionResponse<ContentType>) {
    if (this.props.actionName === 'new') {
      return (allowedChildTypesFromCTD.d.results.length && allowedChildTypesFromCTD.d.results) || []
    }
    return typeResults.length ? typeResults : allowedChildTypesFromCTD.d.results
  }

  private async getAllContentTypes() {
    if (!this.props.repository) {
      throw new Error('You must pass a repository to this control')
    }

    const result = (await this.props.repository.executeAction({
      idOrPath: ConstantContent.PORTAL_ROOT.Id,
      name: 'GetAllContentTypes',
      method: 'GET',
      oDataOptions: {
        select: ['Name', 'DisplayName', 'Icon'],
      },
    })) as ODataCollectionResponse<ContentType>
    if (this.willUnmount) {
      return
    }
    this.setState({
      allCTDs: result.d.results.sort(compare),
      filteredList: result.d.results.sort(compare),
    })
  }

  public handleRemove = (item: GenericContent) => {
    const { items } = this.state
    const index = items.findIndex(i => item.Name === i.Name)
    if (items.length > 1) {
      this.setState({
        items: [...items.slice(0, index), ...items.slice(index + 1)],
      })
    } else {
      this.setState({
        items: this.state.allowedTypesOnCTD,
        removeable: false,
      })
    }
  }

  public handleInputChange = (e: React.ChangeEvent<HTMLTextAreaElement | HTMLInputElement | HTMLSelectElement>) => {
    const term = e.target.value
    this.setState({
      filteredList: this.state.allCTDs.filter(ctd => {
        return ctd.DisplayName && ctd.DisplayName.toLowerCase().includes(term.toLowerCase())
      }),
      inputValue: term,
    })
    if (term.length === 0) {
      this.setState({
        selected: null,
      })
    }
  }

  private handleClickAway() {
    this.setState({ isOpened: false })
  }

  public handleSelect(item: ContentType) {
    this.setState({
      inputValue: item.DisplayName || '',
      isOpened: false,
      isLoading: false,
      selected: item,
    })
  }

  public handleOnClick = () => {
    this.setState({
      isOpened: true,
    })
  }

  public handleAddClick = () => {
    const { items, selected, value } = this.state
    const newValue = selected ? [...value, selected.Name] : value
    if (this.state.removeable) {
      this.setState({
        items: selected ? [...items, selected] : items,
        value: newValue,
        selected: null,
        inputValue: '',
        filteredList: this.state.allCTDs,
      })
    } else {
      this.setState({
        items: selected ? [selected] : [],
        value: selected ? [selected.Name] : [],
        selected: null,
        inputValue: '',
        filteredList: this.state.allCTDs,
        removeable: true,
      })
    }
    this.props.fieldOnChange && this.props.fieldOnChange(this.props.settings.Name, newValue as any)
  }

  public render() {
    switch (this.props.actionName) {
      case 'edit':
      case 'new':
        return (
          <ClickAwayListener onClickAway={this.handleClickAway}>
            <FormControl>
              <FormLabel component={'legend' as 'label'}>{this.props.settings.DisplayName}</FormLabel>
              <List dense={true}>
                {this.state.items.map((item, index) => (
                  <ListItem key={index}>
                    <ListItemIcon style={{ margin: 0 }}>
                      {this.props.renderIcon
                        ? this.props.renderIcon(item.Icon ? item.Icon.toLowerCase() : 'contenttype')
                        : renderIconDefault(
                            item.Icon && typeicons[item.Icon.toLowerCase()]
                              ? typeicons[item.Icon.toLowerCase()]
                              : typeicons.contenttype,
                          )}
                    </ListItemIcon>
                    <ListItemText primary={item.DisplayName} />
                    {this.state.removeable ? (
                      <ListItemSecondaryAction>
                        <IconButton aria-label="Remove" onClick={() => this.handleRemove(item)}>
                          {this.props.renderIcon ? this.props.renderIcon('delete') : renderIconDefault('delete')}
                        </IconButton>
                      </ListItemSecondaryAction>
                    ) : null}
                  </ListItem>
                ))}
              </List>
              <div
                ref={(ref: HTMLDivElement) => ref && this.state.anchorEl !== ref && this.setState({ anchorEl: ref })}
                style={{ position: 'relative' }}>
                <div style={styles.inputContainer as any}>
                  <TextField
                    type="text"
                    onClick={this.handleOnClick}
                    onChange={this.handleInputChange}
                    placeholder={INPUT_PLACEHOLDER}
                    InputProps={{
                      endAdornment: this.state.isLoading ? (
                        <InputAdornment position="end">
                          <CircularProgress size={16} />
                        </InputAdornment>
                      ) : null,
                    }}
                    fullWidth={true}
                    value={this.state.inputValue}
                    style={styles.input}
                  />
                  <IconButton
                    style={styles.button}
                    disabled={this.state.selected && this.state.selected.Name.length > 0 ? false : true}
                    onClick={this.handleAddClick}>
                    {this.props.renderIcon ? this.props.renderIcon('add') : renderIconDefault('add')}
                  </IconButton>
                </div>
                <Paper
                  style={
                    this.state.isOpened
                      ? { ...styles.ddIsOpened, ...(styles.listContainer as any) }
                      : { ...styles.ddIsClosed, ...(styles.listContainer as any) }
                  }>
                  <List>
                    {this.state.filteredList.length > 0 ? (
                      this.state.filteredList.map((item: any) => this.state.getMenuItem(item, this.handleSelect))
                    ) : (
                      <ListItem>No hits</ListItem>
                    )}
                  </List>
                </Paper>
                <FormHelperText>{this.props.settings.Description}</FormHelperText>
              </div>
            </FormControl>
          </ClickAwayListener>
        )
      case 'browse':
      default:
        return (
          <FormControl>
            <FormLabel component={'legend' as 'label'}>{this.props.settings.DisplayName}</FormLabel>
            <List dense={true}>
              {this.state.items.map((item, index) => (
                <ListItem key={index}>
                  <ListItemIcon style={{ margin: 0 }}>
                    {this.props.renderIcon
                      ? this.props.renderIcon(item.Icon ? item.Icon.toLowerCase() : 'contenttype')
                      : renderIconDefault(
                          item.Icon && typeicons[item.Icon.toLowerCase()]
                            ? typeicons[item.Icon.toLowerCase()]
                            : typeicons.contenttype,
                        )}
                  </ListItemIcon>
                  <ListItemText primary={item.DisplayName} />
                </ListItem>
              ))}
            </List>
          </FormControl>
        )
    }
  }
}
