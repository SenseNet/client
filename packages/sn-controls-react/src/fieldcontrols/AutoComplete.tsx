import React from 'react'
import { GenericContent, ReferenceFieldSetting } from '@sensenet/default-content-types'
import { ReferenceField } from '@sensenet/search-react'
import { Query, QueryExpression, QueryOperators } from '@sensenet/query'
import Typography from '@material-ui/core/Typography'
import { changeJScriptValue } from '../helpers'
import { ReactClientFieldSetting } from '.'

/**
 * Represents an autocomplete component
 */
export const AutoComplete: React.FC<ReactClientFieldSetting<ReferenceFieldSetting>> = (props) => {
  const defaultValue =
    (props.fieldValue && (props.fieldValue as any)[0].Id) || changeJScriptValue(props.settings.DefaultValue)
  const fetchItems = async (fetchQuery: Query<GenericContent>) => {
    try {
      if (!props.repository) {
        throw new Error('You must pass a repository to this control')
      }

      new QueryOperators(fetchQuery).and.query((q2) => {
        props.settings.AllowedTypes &&
          props.settings.AllowedTypes.forEach((allowedType, index, array) => {
            new QueryExpression(q2.queryRef).term(`TypeIs:${allowedType}`)
            if (index < array.length - 1) {
              return new QueryOperators(q2.queryRef).or
            }
          })
        return q2
      })

      new QueryOperators(fetchQuery).and.query((q2) => {
        props.settings.SelectionRoots &&
          props.settings.SelectionRoots.forEach((root, index, array) => {
            new QueryExpression(q2.queryRef).inTree(root)
            if (index < array.length - 1) {
              return new QueryOperators(q2.queryRef).or
            }
          })
        return q2
      })

      const result = await props.repository.loadCollection<GenericContent>({
        path: '/Root',
        oDataOptions: {
          query: fetchQuery.toString(),
          select: 'all',
        },
      })
      return result.d.results
    } catch (error) {
      console.error(error.message)
    }
  }

  switch (props.actionName) {
    case 'edit':
    case 'new':
      return (
        <ReferenceField
          fieldSetting={props.settings}
          fieldName={props.settings.Name as any}
          defaultValueIdOrPath={defaultValue}
          onChange={(item) => props.fieldOnChange && props.fieldOnChange(props.settings.Name, item)}
          fetchItems={fetchItems as any}
        />
      )
    case 'browse':
    default: {
      return props.fieldValue ? (
        <div>
          <Typography variant="caption" gutterBottom={true}>
            {props.settings.DisplayName}
          </Typography>
          <Typography variant="body1" gutterBottom={true}>
            {(props.fieldValue as any)[0].DisplayName}
          </Typography>
        </div>
      ) : null
    }
  }
}
