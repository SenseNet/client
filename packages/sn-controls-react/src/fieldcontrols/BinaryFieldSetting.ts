/**
 * @module FieldControls
 *
 */ /** */
import { GenericContent } from '@sensenet/default-content-types'
import { ReactClientFieldSetting } from './ClientFieldSetting'

/**
 * Interface for ShortTextFieldSetting properties
 */
export interface ReactBinaryFieldSetting<T extends GenericContent = GenericContent, K extends keyof T = 'Name'>
  extends ReactClientFieldSetting<T, K> {
  /**
   * Path of the folder where the file should be uploaded
   */
  'data-onChange'?: () => void
  'data-folderPath'?: string
  'data-onUploadSuccess'?: (filename: string, path: string) => void
  'data-innerComponent'?: (filename: string, path: string) => React.Component
}
