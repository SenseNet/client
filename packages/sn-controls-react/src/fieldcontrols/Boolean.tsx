/**
 * @module FieldControls
 */
import { FieldSetting } from '@sensenet/default-content-types'
import React, { useState } from 'react'
import Checkbox from '@material-ui/core/Checkbox'
import FormControl from '@material-ui/core/FormControl'
import FormControlLabel from '@material-ui/core/FormControlLabel'
import FormHelperText from '@material-ui/core/FormHelperText'
import { changeJScriptValue } from '../helpers'
import { ReactClientFieldSetting } from './ClientFieldSetting'
import { renderIconDefault } from './icon'

/**
 * Field control that represents a Boolean field.
 */
export const BooleanComponent: React.FC<ReactClientFieldSetting<FieldSetting>> = (props) => {
  const initialState = props.fieldValue != null ? !!props.fieldValue : !!changeJScriptValue(props.settings.DefaultValue)
  const [value, setValue] = useState(initialState)

  const handleChange = () => {
    setValue(!value)
    props.fieldOnChange && props.fieldOnChange(props.settings.Name, !value)
  }

  switch (props.actionName) {
    case 'edit':
    case 'new':
      return (
        <FormControl required={props.settings.Compulsory} disabled={props.settings.ReadOnly}>
          <FormControlLabel
            name={props.settings.Name}
            control={<Checkbox checked={value} onChange={handleChange} />}
            label={props.settings.DisplayName}
          />
          <FormHelperText>{props.settings.Description}</FormHelperText>
        </FormControl>
      )
    case 'browse':
    default:
      return props.fieldValue != null ? (
        <div style={{ display: 'flex' }}>
          <span>{props.settings.DisplayName}</span>
          {props.renderIcon
            ? props.renderIcon(props.fieldValue ? 'check' : 'not_interested')
            : renderIconDefault(props.fieldValue ? 'check' : 'not_interested')}
        </div>
      ) : null
  }
}
