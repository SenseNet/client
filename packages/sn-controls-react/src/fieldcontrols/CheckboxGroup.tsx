/**
 * @module FieldControls
 */
import React, { useState } from 'react'
import Checkbox from '@material-ui/core/Checkbox'
import FormControl from '@material-ui/core/FormControl'
import FormControlLabel from '@material-ui/core/FormControlLabel'
import FormGroup from '@material-ui/core/FormGroup'
import FormHelperText from '@material-ui/core/FormHelperText'
import FormLabel from '@material-ui/core/FormLabel'
import TextField from '@material-ui/core/TextField'
import { ChoiceFieldSetting } from '@sensenet/default-content-types'
import { ReactClientFieldSetting } from './ClientFieldSetting'

/**
 * Field control that represents a Choice field. Available values will be populated from the FieldSettings.
 */
export const CheckboxGroup: React.FC<ReactClientFieldSetting<ChoiceFieldSetting>> = props => {
  const initialState =
    props.settings.Options &&
    props.settings.Options.map(item =>
      props.fieldValue && Array.isArray(props.fieldValue)
        ? props.fieldValue.some((val: string) => val === item.Value)
          ? { ...item, Selected: true }
          : { ...item, Selected: false }
        : item.Value === props.fieldValue
        ? { ...item, Selected: true }
        : { ...item, Selected: false },
    )
  const [state, setState] = useState(initialState || [])

  const handleChange = (name: string) => (event: React.ChangeEvent<HTMLInputElement>) => {
    const newState = [...state]
    const index = newState.findIndex(item => item.Value === name)
    if (props.settings.AllowMultiple) {
      newState[index].Selected = event.target.checked
    } else {
      newState.forEach(item => (item.Selected = false))
      newState[index].Selected = event.target.checked
    }
    setState(newState)
    props.fieldOnChange && props.fieldOnChange(props.settings.Name, newState.map(item => item.Selected && item.Value))
  }

  switch (props.actionName) {
    case 'edit':
    case 'new':
      return (
        <FormControl
          disabled={props.settings.ReadOnly}
          component={'fieldset' as 'div'}
          required={props.settings.Compulsory}>
          <FormLabel component={'legend' as 'label'}>{props.settings.DisplayName}</FormLabel>
          <FormGroup>
            {state.map(option => {
              return (
                <FormControlLabel
                  key={option.Value}
                  control={
                    <Checkbox checked={option.Selected} onChange={handleChange(option.Value)} value={option.Value} />
                  }
                  label={option.Text}
                />
              )
            })}
          </FormGroup>
          {props.settings.AllowExtraValue ? <TextField placeholder="Extra value" /> : null}
          <FormHelperText>{props.settings.Description}</FormHelperText>
        </FormControl>
      )
    case 'browse':
    default: {
      return props.fieldValue ? (
        <FormControl component={'fieldset' as 'div'}>
          <FormLabel component={'legend' as 'label'}>{props.settings.DisplayName}</FormLabel>
          <FormGroup>
            {Array.isArray(props.fieldValue) ? (
              props.fieldValue.map((val: any, index: number) => (
                <FormControl component={'fieldset' as 'div'} key={index}>
                  <FormControlLabel
                    style={{ marginLeft: 0 }}
                    label={props.settings.Options!.find(item => item.Value === val)!.Text}
                    control={<span />}
                    key={val}
                  />
                </FormControl>
              ))
            ) : (
              <FormControl component={'fieldset' as 'div'}>
                <FormControlLabel
                  style={{ marginLeft: 0 }}
                  label={props.settings.Options!.find(item => item.Value === (props.fieldValue as string))!.Text}
                  control={<span />}
                />
              </FormControl>
            )}
          </FormGroup>
        </FormControl>
      ) : null
    }
  }
}
