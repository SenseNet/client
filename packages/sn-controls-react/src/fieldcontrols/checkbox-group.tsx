/**
 * @module FieldControls
 */
import { deepMerge } from '@sensenet/client-utils'
import { ChoiceFieldSetting } from '@sensenet/default-content-types'
import Checkbox from '@material-ui/core/Checkbox'
import FormControl from '@material-ui/core/FormControl'
import FormControlLabel from '@material-ui/core/FormControlLabel'
import FormGroup from '@material-ui/core/FormGroup'
import FormHelperText from '@material-ui/core/FormHelperText'
import FormLabel from '@material-ui/core/FormLabel'
import TextField from '@material-ui/core/TextField'
import Typography from '@material-ui/core/Typography'
import React, { useState } from 'react'
import { ReactClientFieldSetting } from './client-field-setting'
import { defaultLocalization } from './localization'

/**
 * Field control that represents a Choice field. Available values will be populated from the FieldSettings.
 */
export const CheckboxGroup: React.FC<ReactClientFieldSetting<ChoiceFieldSetting>> = (props) => {
  const localization = deepMerge(defaultLocalization.checkboxGroup, props.localization?.checkboxGroup)

  const initialState =
    props.settings.Options &&
    props.settings.Options.map((item) =>
      props.fieldValue && Array.isArray(props.fieldValue)
        ? props.fieldValue.some((val: string) => val === item.Value)
          ? { ...item, Selected: true }
          : { ...item, Selected: false }
        : item.Value === props.fieldValue
        ? { ...item, Selected: true }
        : { ...item, Selected: false },
    )
  const [state, setState] = useState(initialState || [])

  const handleChange = (name: string) => (event: React.ChangeEvent<HTMLInputElement>) => {
    const newState = [...state]
    const index = newState.findIndex((item) => item.Value === name)
    if (props.settings.AllowMultiple) {
      newState[index].Selected = event.target.checked
    } else {
      newState.forEach((item) => (item.Selected = false))
      newState[index].Selected = event.target.checked
    }
    setState(newState)
    props.fieldOnChange?.(
      props.settings.Name,
      newState.filter((item) => item.Selected).map((item) => item.Value),
    )
  }

  switch (props.actionName) {
    case 'new':
    case 'edit':
      return (
        <FormControl
          disabled={props.settings.ReadOnly}
          component={'fieldset' as 'div'}
          required={props.settings.Compulsory}>
          <FormLabel style={{ transform: 'translate(0, 1.5px) scale(0.75)', transformOrigin: 'top left' }}>
            {props.settings.DisplayName}
          </FormLabel>
          <FormGroup>
            {state.map((option) => {
              return (
                <FormControlLabel
                  key={option.Value}
                  control={
                    <Checkbox checked={option.Selected} onChange={handleChange(option.Value)} value={option.Value} />
                  }
                  label={option.Text}
                />
              )
            })}
          </FormGroup>
          {props.settings.AllowExtraValue ? <TextField placeholder="Extra value" /> : null}
          {!props.hideDescription && <FormHelperText>{props.settings.Description}</FormHelperText>}
        </FormControl>
      )
    case 'browse':
    default: {
      return (
        <>
          <Typography variant="caption" gutterBottom={true}>
            {props.settings.DisplayName}
          </Typography>
          <FormGroup>
            {props.fieldValue ? (
              Array.isArray(props.fieldValue) ? (
                props.fieldValue.length ? (
                  props.fieldValue.map((val: any, index: number) => (
                    <Typography variant="body1" gutterBottom={index === props.fieldValue!.length - 1} key={index}>
                      {props.settings.Options?.find((item) => item.Value === val)?.Text ?? ''}
                    </Typography>
                  ))
                ) : (
                  <Typography variant="body1" gutterBottom={true}>
                    {localization.noValue}
                  </Typography>
                )
              ) : (
                <Typography variant="body1" gutterBottom={true}>
                  {props.settings.Options?.find((item) => item.Value === (props.fieldValue as string))?.Text ?? ''}
                </Typography>
              )
            ) : (
              <Typography variant="body1" gutterBottom={true}>
                {localization.noValue}
              </Typography>
            )}
          </FormGroup>
        </>
      )
    }
  }
}
