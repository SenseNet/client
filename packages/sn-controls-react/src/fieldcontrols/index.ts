/**
 * @module FieldControls
 * @preferred
 * @description This module contains controls that are used to display or edit specific fields of a specified Content.
 */

export * from './Number'
export * from './CheckboxGroup'
export * from './DatePicker'
export * from './TimePicker'
export * from './DropDownList'
export * from './RadioButtonGroup'
export * from './ShortText'
export * from './TagsInput'
export * from './Textarea'
export * from './RichTextEditor/RichTextEditor'
export * from './Name'
export * from './Password'
export * from './FileName'
export * from './AutoComplete'
export * from './FileUpload'
export * from './Avatar/Avatar'
export * from './ReferenceGrid/ReferenceGrid'
export * from './Boolean'
export * from './EmptyFieldControl'
export * from './ColorPicker'
export * from './AllowedChildTypes'
export * from './ClientFieldSetting'
export * from './switcher'
