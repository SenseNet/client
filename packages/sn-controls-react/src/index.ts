/**
 * @module sn-controls-react
 * @description General classes, modules and methods for sn-controls-react package
 */

export * from './ReactControlMapper'
export * from './fieldcontrols'
export * from './viewcontrols'
