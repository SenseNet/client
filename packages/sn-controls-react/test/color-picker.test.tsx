import TextField from '@material-ui/core/TextField'
import Typography from '@material-ui/core/Typography'
import { shallow } from 'enzyme'
import React from 'react'
import { SketchPicker } from 'react-color'
import { ColorPicker, defaultLocalization } from '../src/fieldcontrols'

describe('Color field control', () => {
  const defaultSettings = {
    Name: 'Color',
    FieldClassName: 'SenseNet.ContentRepository.Fields.ColorField',
    DisplayName: 'Color',
    Description: 'Color',
    Type: 'ColorFieldSetting',
  }
  const value = '#016d9e'
  describe('in browse view', () => {
    it('should show the displayname and fieldValue when fieldValue is provided', () => {
      const wrapper = shallow(<ColorPicker fieldValue={value} actionName="browse" settings={defaultSettings} />)
      const labelElement = wrapper.find(Typography).first()
      const valueElement = wrapper.find(Typography).at(1)
      expect(labelElement.props().children).toEqual(defaultSettings.DisplayName)
      expect(valueElement.props().children).toBe(value)
      expect(wrapper).toMatchSnapshot()
    })
    it('should show no value message when field value is not provided', () => {
      const wrapper = shallow(<ColorPicker actionName="browse" settings={defaultSettings} />)
      expect(wrapper.find(Typography).last().text()).toBe(defaultLocalization.colorPicker.noValue)
    })
  })
  describe('in edit/new view', () => {
    it('should set all the props', () => {
      const wrapper = shallow(
        <ColorPicker
          fieldValue={value}
          actionName="edit"
          settings={{
            ...defaultSettings,
            ReadOnly: true,
            Compulsory: true,
          }}
        />,
      )
      expect(wrapper.find(TextField).prop('value')).toBe(value)
      expect(wrapper.find(TextField).prop('name')).toBe(defaultSettings.Name)
      expect(wrapper.find(TextField).prop('id')).toBe(defaultSettings.Name)
      expect(wrapper.find(TextField).prop('label')).toBe(defaultSettings.DisplayName)
      expect(wrapper.find(TextField).prop('placeholder')).toBe(defaultSettings.DisplayName)
      expect(wrapper.find(TextField).prop('required')).toBeTruthy()
      expect(wrapper.find(TextField).prop('disabled')).toBeTruthy()
      expect(wrapper.find(TextField).prop('helperText')).toBe(defaultSettings.Description)
      expect(wrapper).toMatchSnapshot()
    })

    it('should set default value', () => {
      const wrapper = shallow(
        <ColorPicker
          actionName="new"
          settings={{
            ...defaultSettings,
            DefaultValue: 'defaultValue',
          }}
        />,
      )

      expect(wrapper.find(TextField).prop('value')).toBe('defaultValue')
    })
    it('should open color picker when input is clicked', () => {
      const wrapper = shallow(<ColorPicker actionName="edit" settings={defaultSettings} />)
      wrapper.find(TextField).simulate('click')
      expect(wrapper.exists(SketchPicker)).toBeTruthy()
    })
    it('should handle color change', () => {
      const fieldChange = jest.fn()
      const wrapper = shallow(<ColorPicker actionName="edit" fieldOnChange={fieldChange} settings={defaultSettings} />)
      wrapper.find(TextField).simulate('click')
      const onChange: any | undefined = wrapper.find(SketchPicker).prop('onChangeComplete')
      onChange?.({ hex: '#111111' })
      expect(fieldChange).toBeCalled()
    })
  })
})
