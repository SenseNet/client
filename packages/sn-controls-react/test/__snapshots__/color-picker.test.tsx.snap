// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Color field control in browse view should show the displayname and fieldValue when fieldValue is provided 1`] = `
<Fragment>
  <WithStyles(ForwardRef(Typography))
    gutterBottom={true}
    variant="caption"
  >
    Color
  </WithStyles(ForwardRef(Typography))>
  <div
    style={
      Object {
        "display": "flex",
      }
    }
  >
    <WithStyles(ForwardRef(Icon))
      style={
        Object {
          "color": "#016d9e",
        }
      }
    >
      lens
    </WithStyles(ForwardRef(Icon))>
    <WithStyles(ForwardRef(Typography))
      gutterBottom={true}
      style={
        Object {
          "marginLeft": "0.5rem",
        }
      }
      variant="body1"
    >
      #016d9e
    </WithStyles(ForwardRef(Typography))>
  </div>
</Fragment>
`;

exports[`Color field control in edit/new view should set all the props 1`] = `
<WithStyles(ForwardRef(FormControl))>
  <WithStyles(ForwardRef(TextField))
    InputProps={
      Object {
        "startAdornment": <WithStyles(ForwardRef(InputAdornment))
          position="start"
        >
          <WithStyles(ForwardRef(Icon))
            style={
              Object {
                "color": "#016d9e",
              }
            }
          >
            lens
          </WithStyles(ForwardRef(Icon))>
        </WithStyles(ForwardRef(InputAdornment))>,
      }
    }
    defaultValue="defaultValue"
    disabled={true}
    helperText="Color"
    id="Color"
    label="Color"
    name="Color"
    onClick={[Function]}
    placeholder="Color"
    required={true}
    type="text"
    value="#016d9e"
  />
</WithStyles(ForwardRef(FormControl))>
`;
